cloudbuild/python-scripts/insert_dynamic_partitions.py

```diff
@@ -0,0 +1,146 @@
+
+    # Providing the values.
+    project_name = args.project_name
+    dataset_name = args.dataset_name
+    table_name = args.table_name
+    now_timestamp = args.now_timestamp
+    now_timestamp = datetime.datetime.strptime(
+        now_timestamp, '%Y-%m-%d'
+    ).astimezone(datetime.timezone.utc)
+    refresh_interval = int(args.refresh_interval)
+
+    # Set the partitioned table.
+    table_id = f'{project_name}.{dataset_name}.{table_name}'
+
+    # Now add the partitions to the table.
+    # Hardcoded schema. Needs to be same as that in the pre-created table.
+    simple_avro_schema_fields_string = (
+        '"fields": [{"name": "name", "type": "string"},{"name": "number",'
+        '"type": "long"},{"name" : "ts", "type" : {"type" :'
+        '"long","logicalType": "timestamp-micros"}}]'
+    )
+    simple_avro_schema_string = (
+        '{"namespace": "project.dataset","type": "record","name":'
+        ' "table","doc": "Avro Schema for project.dataset.table",'
+        + simple_avro_schema_fields_string
+        + '}'
+    )
+
+    # hardcoded for e2e test.
+    # partitions[i] * number_of_rows_per_partition are inserted per phase.
+    partitions = [2, 1, 2]
+    # Insert 1000 - 3000 rows per partition.
+    # So, in a read up to 6000 new rows are read.
+    number_of_rows_per_partition = random.randint(1, 3) * 1000
+    number_of_threads = 10
+    number_of_rows_per_thread = int(
+        number_of_rows_per_partition / number_of_threads
+    )
+
+    avro_file_local = 'mockData.avro'
+    table_creation_utils = utils.TableCreationUtils(
+        simple_avro_schema_string,
+        number_of_rows_per_thread,
+        table_id,
+    )
+
+    # Insert in phases.
+    prev_partitions_offset = 0
+    for number_of_partitions in partitions:
+        start_time = time.time()
+        prev_partitions_offset += 1
+        # Wait for the connector to read previously inserted rows.
+        wait()
+
+        # This is a phase of insertion.
+        for partition_number in range(number_of_partitions):
+            threads = list()
+            # Insert via concurrent threads.
+            for thread_number in range(number_of_threads):
+                avro_file_local_identifier = avro_file_local.replace(
+                    '.', '_' + str(thread_number) + '.'
+                )
+                x = threading.Thread(
+                    target=table_creation_utils.avro_to_bq_with_cleanup,
+                    kwargs={
+                        'avro_file_local_identifier': avro_file_local_identifier,
+                        'partition_number': partition_number + prev_partitions_offset,
+                        'current_timestamp': now_timestamp,
+                    },
+                )
+                threads.append(x)
+                x.start()
+            for _, thread in enumerate(threads):
+                thread.join()
+
+        time_elapsed = time.time() - start_time
+        prev_partitions_offset += number_of_partitions
```

Comment:
Why is `prev_partitions_offset` being incremented multiple times in the same iteration?

Reply:
Within the same iteration we are adding rows spread amongst 1 or more partitions.
so that at a new read, we make sure that multiple partitions are being read from.

Reply:
Let's take an example.

First iteration:
```
    for number_of_partitions in partitions:   # number_of_partitions is 2
        ...
        prev_partitions_offset += 1   # prev_partitions_offset is 1
        ...
        # called avro_to_bq_with_cleanup with partition_number as 1
        # called avro_to_bq_with_cleanup with partition_number as 2
        ...
        prev_partitions_offset += number_of_partitions   # prev_partitions_offset is 3
        ...
```

Second iteration:
```
    for number_of_partitions in partitions:   # number_of_partitions is 1
        ...
        prev_partitions_offset += 1   # prev_partitions_offset is 4
        ...
        # called avro_to_bq_with_cleanup with partition_number as 4
        ...
        prev_partitions_offset += number_of_partitions   # prev_partitions_offset is 5
        ...
```

Third iteration:
```
    for number_of_partitions in partitions:   # number_of_partitions is 2
        ...
        prev_partitions_offset += 1   # prev_partitions_offset is 6
        ...
        # called avro_to_bq_with_cleanup with partition_number as 6
        # called avro_to_bq_with_cleanup with partition_number as 7
        ...
        prev_partitions_offset += number_of_partitions   # prev_partitions_offset is 8
        ...
```

So, we've skipped partition offsets 3 and 5. If that is intentional, then why?

Reply:
Yeah, this is correct.
So to maintain time consistency, I took time at UTC which would be 18:30 hrs.
So (18:30 + 2 = 20:30) to (18:30 + 3 = 21:30) will generate values in partitions 20hrs, 21 hrs.

So in next phase if we generate for 18:30 + 3 - 18:30 + 4, the partitions will clash. 

I think this is getting too confusing. 
I'll fix this.

Reply:
This is fixed now.